migrations/20240118090944_create_function_tx_join_group_tx_msg_and_coin_transfer.ts

```diff
@@ -0,0 +1,101 @@
+import { Knex } from 'knex';
+
+export async function up(knex: Knex): Promise<void> {
+  await knex.schema.raw(
+    `CREATE OR REPLACE function transaction_join_group_transaction_message(
+      senderAddress character varying, 
+      typesIn character varying[], 
+      typesNotIn character varying[],
+      limitValue int)  
+    RETURNS SETOF transaction
+    LANGUAGE plpgsql  
+    STABLE PARALLEL SAFE
+    AS  
+    $$  
+    begin
+      if typesIn is null then 
+        if typesNotIn is null then
+          return query (select transaction.* from transaction right join (
+            select tx_id as txid from transaction_message
+            where transaction_message.sender = senderAddress
+            group by tx_id order by tx_id
+            limit limitValue
+              ) a on a.txid = transaction.id);
+        else
+          return query (select transaction.* from transaction right join (
+            select tx_id as txid from transaction_message
+            where transaction_message.sender = senderAddress
+            and type <> ALL(typesNotIn)
+            group by tx_id order by tx_id
+            limit limitValue
+          ) a on a.txid = transaction.id);
+        end if;
+      else
+        return query (select transaction.* from transaction right join (
+          select tx_id as txid from transaction_message
+          where transaction_message.sender = senderAddress
+          and type = ANY(typesIn)
+          group by tx_id order by tx_id
+          limit limitValue
+        ) a on a.txid = transaction.id);
+      end if;
+    end;
+    $$;`
+  );
+  await knex.schema.raw(
+    `CREATE OR REPLACE FUNCTION transaction_join_group_coin_transfer(
+      addressFrom character varying, 
+      addressTo character varying,
+      typesIn character varying[], 
+      typesNotIn character varying[],
+      limitValue integer)
+    RETURNS SETOF transaction
+    LANGUAGE plpgsql
+    STABLE PARALLEL SAFE
+    AS $$
+    begin
+      if typesIn is null then
+        if typesNotIn is null then
```

Comment:
như vậy điều kiện ở đây là:
- ko có cả in và not in
- có not in
- có in
thế thôi đúng ko?

Reply:
vâng đúng rồi